atlas: draw selection colors as background/foreground instead of alpha overlay

DHowett opened the pull request:
With the merge of #17638, selections are now accumulated early in the
rendering process. This allows Atlas, which currently makes decisions
about cell foreground/background at the time of text rendering,
awareness of the selection ranges *before* text rendering begins.

As a result, we can now paint the selection into the background and
foreground bitmaps. We no longer need to overlay a rectangle, or series
of rectangles, on top of the rendering surface and alpha blend the
selection color onto the final image.

As a reminder, "alpha selection" was always a stopgap because we didn't
have durable per-cell foreground and background customization in the
original DxEngine.

Selection foregrounds are not customizable, and will be chosen using the
same color distancing algorithm as the cursor. We can make them
customizable "easily" (once we figure out the schema for it) for #3580.

`ATLAS_DEBUG_SHOW_DIRTY` was using the `Selection` shading type to draw
colored regions. I didn't want to break that, so I elected to rename the
`Selection` shading type to `FilledRect` and keep its value. It helps
that the shader didn't have any special treatment for
`SHADING_TYPE_SELECTION`.

This fixes the entire category of issues created by selection being an
80%-opacity white rectangle. However, given that it changes the imputed
colors of the text it will reveal `SGR 8` concealed/hidden characters.

Refs #17355
Refs #14859
Refs #11181
Refs #8716
Refs #4971
Closes #3561

DHowett commented on src/renderer/atlas/AtlasEngine.cpp:
```diff
@@ -411,9 +413,10 @@ try
+        // Return early if we couldn't paint the whole region (either this was not the last row, or
+        // it was the last row but the highlight ends outside of our x range.)
+        // We will resume from here in the next call.
+        if (!isFinalRow || hiEnd.x /*inclusive*/ >= x2 /*exclusive*/)
```
This is the thing it took hours to fix lol

tusharsnx replied:
Wait, Ahh, hmm. Thanks for fixing this 🥲😃

DHowett commented on src/renderer/atlas/shader_common.hlsl:
```diff
@@ -16,7 +16,7 @@
 #define SHADING_TYPE_SOLID_LINE         8
 #define SHADING_TYPE_CURSOR             9
-#define SHADING_TYPE_SELECTION          10
+#define SHADING_TYPE_FILLED_RECT       10
```
better name would be A-OK with me

DHowett commented:
Here's a quick demo video of selection colors, and how they change while I change color schemes.


https://github.com/user-attachments/assets/ad2261a3-f577-4cb2-b2af-7af4b6fedb6b

lhecker commented on src/renderer/atlas/AtlasEngine.api.cpp:
```diff
@@ -424,12 +425,14 @@ void AtlasEngine::SetRetroTerminalEffect(bool enable) noexcept
-        _api.s.write()->misc.write()->selectionColor = selectionColor;
+        auto misc = _api.s.write()->misc.write();
+        misc->selectionColor = selectionColor;
+        misc->selectionForeground = 0xff000000 | ColorFix::GetPerceivableColor(color, color, 0.5f * 0.5f);
```
Hold up, this has `color` twice.

DHowett replied:
Oh, no, I didn't misread. I pulled this from BackendD3D `_drawCursorForegroundSlowPath`

```
    auto color = c.foreground == 0xffffffff ? it.color ^ 0xffffff : c.foreground;
    color = ColorFix::GetPerceivableColor(color, c.background, 0.5f * 0.5f);
```

It uses a foreground and a background, whereas we don't _have_ a foreground color to speak of yet.

DHowett commented:
@tusharsnx let me tell you, I spent five hours with a whiteboard and rewrote the whole thing, and it was a seriously humbling experience. I fixed the issue I was seeing, and then I decided to just keep everything that was already there.

I couldn't--and still can't!--wrap my head around how to simplify it given that it detects overlaps of a coordinate span with a coordinate span. In my rewrite, I ended up linearizing every coordinate (y*w+x) instead.

I'm still mad at myself for it. 😁

DHowett commented:
/azp run
